tester/pkg/kops.go

```diff
@@ -0,0 +1,258 @@
+package pkg
+
+import (
+	"fmt"
+	"io/ioutil"
+	"log"
+	"net/http"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"runtime"
+	"time"
+)
+
+const (
+	kopsStateFile = "s3://k8s-kops-csi-e2e"
+)
+
+type FuncStep struct {
+	f func() error
+}
+
+func (s *FuncStep) Run() error {
+	return s.f()
+}
+
+func EmptyStep() Step {
+	return &FuncStep{func() error {
+		return nil
+	}}
+}
+
+type KopsClusterCreator struct {
+	// TestId is used as cluster name
+	TestId string
+
+	// Kops represents the configuration to create kops cluster
+	Kops *KopsCluster
+
+	// directory where tempory data is saved
+	TestDir string
+
+	// The path to Kops executable
+	KopsBinaryPath string
+}
+
+func NewKopsClusterCreator(kops *KopsCluster, dir string, testId string) *KopsClusterCreator {
+	binaryFilePath := filepath.Join(dir, "kops")
+	return &KopsClusterCreator{
+		Kops:           kops,
+		TestDir:        dir,
+		TestId:         testId,
+		KopsBinaryPath: binaryFilePath,
+	}
+}
+
+func (c *KopsClusterCreator) Init() (Step, error) {
+	f := func() error {
+		_, err := os.Stat(c.TestDir)
+		if os.IsNotExist(err) {
+			err := os.Mkdir(c.TestDir, 0777)
+			if err != nil {
+				return err
+			}
+		}
+
+		_, err = os.Stat(c.KopsBinaryPath)
+		if os.IsNotExist(err) {
+			return c.downloadKops()
+		}
+
+		return nil
+	}
+
+	return &FuncStep{f}, nil
+}
+
+func (c *KopsClusterCreator) Up() (Step, error) {
+	f := func() error {
+		// create cluster
+		err := c.createCluster()
+		if err != nil {
+			return err
+		}
+
+		c.waitForCreation()
+
+		// wait for cluster creation to success
+		// or return err if failed
```

Comment:
TODO to have a timeout from waitForCreation

Reply:
added timeout